rustworkx-core/src/dag_algo.rs

```diff
@@ -333,6 +337,154 @@ where
 }
 
+/// Return an iterator of graph layers
+///
+/// A layer is a subgraph whose nodes are disjoint, i.e.,
+/// a layer has depth 1. The layers are constructed using a greedy algorithm.
+///
+/// Arguments:
+///
+/// * `graph` - The graph to get the layers from
+/// * `first_layer` - A list of node ids for the first layer. This
+///     will be the first layer in the output
+///
+/// Will `panic!` if a provided node is not in the graph.
+/// ```
+/// use rustworkx_core::petgraph::prelude::*;
+/// use rustworkx_core::dag_algo::layers;
+/// use rustworkx_core::dictmap::*;
+///
+/// let edge_list = vec![
+///  (0, 1),
+///  (1, 2),
+///  (2, 3),
+///  (3, 4),
+/// ];
+///
+/// let graph = DiGraph::<u32, u32>::from_edges(&edge_list);
+/// let layers: Vec<Vec<NodeIndex>> = layers(&graph, vec![0.into(),]).map(|layer| layer.unwrap()).collect();
+/// let expected_layers: Vec<Vec<NodeIndex>> = vec![
+///     vec![0.into(),],
+///     vec![1.into(),],
+///     vec![2.into(),],
+///     vec![3.into(),],
+///     vec![4.into()]
+/// ];
+/// assert_eq!(layers, expected_layers)
+/// ```
+pub fn layers<G>(
+    graph: G,
+    first_layer: Vec<G::NodeId>,
+) -> impl Iterator<Item = Result<Vec<G::NodeId>, LayersError>>
+where
+    G: NodeIndexable // Used in from_index and to_index.
+        + IntoNodeIdentifiers // Used for .node_identifiers
+        + IntoNeighborsDirected // Used for .neighbors_directed
+        + IntoEdgesDirected, // Used for .edged_directed
+    <G as GraphBase>::NodeId: Debug + Copy + Eq + Hash,
+{
+    LayersIter {
+        graph,
+        cur_layer: first_layer,
+        next_layer: vec![],
+        predecessor_count: HashMap::new(),
+        first_iter: true,
+        cycle_check: HashSet::default(),
+    }
+}
+
+#[derive(Debug, Clone)]
+struct LayersIter<G, N> {
+    graph: G,
+    cur_layer: Vec<N>,
+    next_layer: Vec<N>,
+    predecessor_count: HashMap<N, usize>,
+    first_iter: bool,
+    cycle_check: HashSet<N>, // TODO: Figure out why some cycles cannot be detected
+}
+
+impl<G, N> Iterator for LayersIter<G, N>
+where
+    G: NodeIndexable // Used in from_index and to_index.
+        + IntoNodeIdentifiers // Used for .node_identifiers
+        + IntoNeighborsDirected // Used for .neighbors_directed
+        + IntoEdgesDirected // Used for .edged_directed
+        + GraphBase<NodeId = N>,
+    N: Debug + Copy + Eq + Hash,
+{
+    type Item = Result<Vec<N>, LayersError>;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.first_iter {
+            self.first_iter = false;
+            for node in &self.cur_layer {
+                if self.graph.to_index(*node) >= self.graph.node_bound() {
+                    panic!("Node {:#?} is not present in the graph.", node);
+                }
+                if self.cycle_check.contains(node) {
+                    return Some(Err(LayersError(
+                        "An invalid first layer was provided.".to_owned(),
+                    )));
```

Comment:
Is it possible to have a cycle here? Or are you protecting against duplicate entries in `first_layer`. If it's duplicates we should put that in the error message.